Intermittent debug(T) - org.mozilla.fenix.tabstray.NavigationInteractorTest - WHEN request desktop menu item is toggled On THEN desktop site is requested for the session

### Firebase Test Run:
https://treeherder.mozilla.org/logviewer?job_id=337798449&repo=fenix
### Stacktrace:
```
[task 2021-04-26T23:40:37.722Z]   TEST: WHEN request desktop menu item is toggled On THEN desktop site is requested for the session
[task 2021-04-26T23:40:37.722Z]     [Robolectric] org.mozilla.fenix.components.toolbar.DefaultBrowserToolbarMenuControllerTest.WHEN request desktop menu item is toggled On THEN desktop site is requested for the session: sdk=28; resources=BINARY
[task 2021-04-26T23:40:37.822Z]     I/MonitoringInstr: Instrumentation started!
[task 2021-04-26T23:40:37.822Z]     I/glean/Dispatchers: Task queued for execution and delayed until flushed
[task 2021-04-26T23:40:37.822Z]     I/glean/Dispatchers: Task queued for execution and delayed until flushed
[task 2021-04-26T23:40:38.022Z]   FAILURE
[task 2021-04-26T23:40:38.022Z] 
[task 2021-04-26T23:40:38.023Z] java.lang.AssertionError: Verification failed: call 2 of 2: AddBookmarksUseCase(child of #168#172).invoke(any(), any(), any(), any())) was not called
[task 2021-04-26T23:40:38.023Z] 	at io.mockk.impl.recording.states.VerifyingState.failIfNotPassed(VerifyingState.kt:66)
[task 2021-04-26T23:40:38.023Z] 	at io.mockk.impl.recording.states.VerifyingState.recordingDone(VerifyingState.kt:42)
[task 2021-04-26T23:40:38.024Z] 	at io.mockk.impl.recording.CommonCallRecorder.done(CommonCallRecorder.kt:47)
[task 2021-04-26T23:40:38.024Z] 	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:60)
[task 2021-04-26T23:40:38.024Z] 	at io.mockk.impl.eval.VerifyBlockEvaluator.verify(VerifyBlockEvaluator.kt:30)
[task 2021-04-26T23:40:38.024Z] 	at io.mockk.MockKDsl.internalCoVerify(API.kt:143)
[task 2021-04-26T23:40:38.024Z] 	at io.mockk.MockKKt.coVerify(MockK.kt:169)
[task 2021-04-26T23:40:38.024Z] 	at io.mockk.MockKKt.coVerify$default(MockK.kt:166)
[task 2021-04-26T23:40:38.024Z] 	at org.mozilla.fenix.tabstray.NavigationInteractorTest$onBookmarkTabs calls navigation on DefaultNavigationInteractor$1.invokeSuspend(NavigationInteractorTest.kt:213)
[task 2021-04-26T23:40:38.024Z] 	at org.mozilla.fenix.tabstray.NavigationInteractorTest$onBookmarkTabs calls navigation on DefaultNavigationInteractor$1.invoke(NavigationInteractorTest.kt)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.test.TestBuildersKt$runBlockingTest$deferred$1.invokeSuspend(TestBuilders.kt:50)
[task 2021-04-26T23:40:38.025Z] 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.test.TestCoroutineDispatcher.dispatch(TestCoroutineDispatcher.kt:50)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:305)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:27)
[task 2021-04-26T23:40:38.025Z] 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:109)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:158)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.async(Builders.common.kt:91)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.BuildersKt.async(Unknown Source)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.async$default(Builders.common.kt:84)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.BuildersKt.async$default(Unknown Source)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.test.TestBuildersKt.runBlockingTest(TestBuilders.kt:49)
[task 2021-04-26T23:40:38.026Z] 	at kotlinx.coroutines.test.TestBuildersKt.runBlockingTest$default(TestBuilders.kt:45)
[task 2021-04-26T23:40:38.026Z] 	at org.mozilla.fenix.tabstray.NavigationInteractorTest.onBookmarkTabs calls navigation on DefaultNavigationInteractor(NavigationInteractorTest.kt:211)
[task 2021-04-26T23:40:38.026Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-04-26T23:40:38.027Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-04-26T23:40:38.027Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-04-26T23:40:38.027Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-04-26T23:40:38.027Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[task 2021-04-26T23:40:38.027Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[task 2021-04-26T23:40:38.027Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[task 2021-04-26T23:40:38.027Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[task 2021-04-26T23:40:38.027Z] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
[task 2021-04-26T23:40:38.028Z] 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
[task 2021-04-26T23:40:38.029Z] 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
[task 2021-04-26T23:40:38.029Z] 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
[task 2021-04-26T23:40:38.029Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
[task 2021-04-26T23:40:38.029Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
[task 2021-04-26T23:40:38.029Z] 	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
[task 2021-04-26T23:40:38.029Z] 	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
[task 2021-04-26T23:40:38.029Z] 	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
[task 2021-04-26T23:40:38.029Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-04-26T23:40:38.029Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-04-26T23:40:38.030Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-04-26T23:40:38.030Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-04-26T23:40:38.030Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
[task 2021-04-26T23:40:38.030Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[task 2021-04-26T23:40:38.030Z] 	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
[task 2021-04-26T23:40:38.030Z] 	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
[task 2021-04-26T23:40:38.030Z] 	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
[task 2021-04-26T23:40:38.030Z] 	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
[task 2021-04-26T23:40:38.030Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2021-04-26T23:40:38.030Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2021-04-26T23:40:38.031Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2021-04-26T23:40:38.031Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
[task 2021-04-26T23:40:38.031Z] 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
[task 2021-04-26T23:40:38.032Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[task 2021-04-26T23:40:38.032Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[task 2021-04-26T23:40:38.032Z] 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
[task 2021-04-26T23:40:38.032Z] 	at java.lang.Thread.run(Thread.java:748)
```
### Build: 
https://github.com/mozilla-mobile/fenix/commit/8070a326b7c9ff626fa854ce9250143d49281412

--- a/app/src/main/java/org/mozilla/fenix/tabstray/NavigationInteractor.kt
+++ b/app/src/main/java/org/mozilla/fenix/tabstray/NavigationInteractor.kt
@@ -7,7 +7,6 @@ package org.mozilla.fenix.tabstray
 import android.content.Context
 import androidx.navigation.NavController
 import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.launch
 import mozilla.components.browser.state.selector.getNormalOrPrivateTabs
@@ -25,6 +24,7 @@ import org.mozilla.fenix.components.metrics.MetricController
 import org.mozilla.fenix.ext.navigateBlockingForAsyncNavGraph
 import org.mozilla.fenix.home.HomeFragment
 import org.mozilla.fenix.tabstray.ext.getTabSessionState
+import kotlin.coroutines.CoroutineContext
 
 /**
  * An interactor that helps with navigating to different parts of the app from the tabs tray.
@@ -91,7 +91,8 @@ class DefaultNavigationInteractor(
     private val bookmarksUseCase: BookmarksUseCase,
     private val tabsTrayStore: TabsTrayStore,
     private val collectionStorage: TabCollectionStorage,
-    private val accountManager: FxaAccountManager
+    private val accountManager: FxaAccountManager,
+    private val ioDispatcher: CoroutineContext
 ) : NavigationInteractor {
 
     override fun onTabTrayDismissed() {
@@ -181,8 +182,9 @@ class DefaultNavigationInteractor(
     override fun onSaveToBookmarks(tabs: Collection<Tab>) {
         tabs.forEach { tab ->
             // We don't combine the context with lifecycleScope so that our jobs are not cancelled
-            // if we leave the fragment, i.e. we still want the bookmarks to be added.
-            CoroutineScope(Dispatchers.IO).launch {
+            // if we leave the fragment, i.e. we still want the bookmarks to be added if the
+            // tabs tray closes before the job is done.
+            CoroutineScope(ioDispatcher).launch {
                 bookmarksUseCase.addBookmark(tab.url, tab.title)
             }
         }

--- a/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt
+++ b/app/src/main/java/org/mozilla/fenix/tabstray/TabsTrayFragment.kt
@@ -119,7 +119,8 @@ class TabsTrayFragment : AppCompatDialogFragment(), TabsTrayInteractor {
                 dismissTabTrayAndNavigateHome = ::dismissTabsTrayAndNavigateHome,
                 bookmarksUseCase = requireComponents.useCases.bookmarksUseCases,
                 collectionStorage = requireComponents.core.tabCollectionStorage,
-                accountManager = requireComponents.backgroundServices.accountManager
+                accountManager = requireComponents.backgroundServices.accountManager,
+                ioDispatcher = Dispatchers.IO
             )
 
         tabsTrayController = DefaultTabsTrayController(

--- a/app/src/test/java/org/mozilla/fenix/tabstray/NavigationInteractorTest.kt
+++ b/app/src/test/java/org/mozilla/fenix/tabstray/NavigationInteractorTest.kt
@@ -14,6 +14,7 @@ import io.mockk.verify
 import io.mockk.mockkStatic
 import io.mockk.unmockkStatic
 import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
 import kotlinx.coroutines.test.runBlockingTest
 import mozilla.components.browser.state.state.BrowserState
 import mozilla.components.browser.state.state.TabSessionState
@@ -21,6 +22,7 @@ import mozilla.components.browser.state.state.createTab as createStateTab
 import mozilla.components.browser.state.store.BrowserStore
 import mozilla.components.concept.tabstray.Tab
 import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.support.test.rule.MainCoroutineRule
 import org.junit.Assert.assertTrue
 import org.junit.Before
 import org.junit.Rule
@@ -34,6 +36,7 @@ import org.mozilla.fenix.components.metrics.Event
 import org.mozilla.fenix.components.metrics.MetricController
 import org.mozilla.fenix.tabstray.browser.createTab as createTrayTab
 
+@ExperimentalCoroutinesApi
 class NavigationInteractorTest {
     private lateinit var store: BrowserStore
     private lateinit var tabsTrayStore: TabsTrayStore
@@ -48,8 +51,13 @@ class NavigationInteractorTest {
     private val collectionStorage: TabCollectionStorage = mockk(relaxed = true)
     private val accountManager: FxaAccountManager = mockk(relaxed = true)
 
+    private val testDispatcher = TestCoroutineDispatcher()
+
     @get:Rule
     val disableNavGraphProviderAssertionRule = DisableNavGraphProviderAssertionRule()
+
+    @get:Rule
+    val coroutinesTestRule = MainCoroutineRule(testDispatcher)
 
     @Before
     fun setup() {
@@ -65,7 +73,8 @@ class NavigationInteractorTest {
             bookmarksUseCase,
             tabsTrayStore,
             collectionStorage,
-            accountManager
+            accountManager,
+            testDispatcher
         )
     }
 
@@ -206,9 +215,21 @@ class NavigationInteractorTest {
         unmockkStatic("org.mozilla.fenix.collections.CollectionsDialogKt")
     }
 
-    @OptIn(ExperimentalCoroutinesApi::class)
     @Test
     fun `onBookmarkTabs calls navigation on DefaultNavigationInteractor`() = runBlockingTest {
+        navigationInteractor = DefaultNavigationInteractor(
+            context,
+            store,
+            navController,
+            metrics,
+            dismissTabTray,
+            dismissTabTrayAndNavigateHome,
+            bookmarksUseCase,
+            tabsTrayStore,
+            collectionStorage,
+            accountManager,
+            coroutineContext
+        )
         navigationInteractor.onSaveToBookmarks(listOf(createTrayTab()))
         coVerify(exactly = 1) { bookmarksUseCase.addBookmark(any(), any(), any()) }
     }

